Add rust v3 (#1)
canonical/ubuntu-mir: README.md

TODO-B: - Team is not yet, but will subscribe to the package before promotion

RULE: - Responsibilities implied by static builds promoted to main, which is
RULE:   not a recommended but a common case with golang and rust packages.
RULE:   - the security team will track CVEs for all vendored/embedded sources in main
RULE:   - the security team will provide updates to main for all `golang-*-dev`
RULE:     packages
RULE:   - the security team will provide updates to main for non-vendored
RULE:     dependencies as per normal procedures (including e.g.,
RULE:     sponsoring/coordinating uploads from teams/upstream projects, etc)
RULE:   - the security team will perform no-change-rebuilds for all packages

> RULE:     listing an CVE-fixed package as Built-Using and coordinate testing

Reviewer: FYI, dh-golang has switched to using `Static-Built-Using:` for reporting the golang-*-dev packages that are used as build time dependencies; this has landed [in kinetic](https://launchpad.net/ubuntu/+source/dh-golang/1.55) and various golang leaf packages are starting to pick it up through rebuilds.

Author: Thanks @stevebeattie - that is interesting - from your need to track content that is contained - would that new `Static-Built-Using:` be equal or better to the former ask to get go.mod/go.sum/cargo.lock? So should we prefer to go towards asking for them instead - which due to the similarity with Debian should be easier to maintain.

Such usually hold packages - do you happen to know if/how vendored dependencies are listed in there?

Author: For reference, this was brought into dh-golang via
- https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=921284
- https://salsa.debian.org/go-team/packages/dh-golang/-/commit/055ddf971b73fb20bf306d8faf7e4a31eb1b3928

dh-cargo has no such change and actually not any recent change, the current version is 7 months old.

Author: We discussed about this, not only does it not exist in the rust tools - it is only for the packaged dependencies and the whole "what if vendored" discussion is due to Ubuntu allowing special cases with vendoring - and those would not be covered by this.

RULE:       c) The need to backport the fix in the dependency, as it would imply
RULE:          requiring a newer toolchain to be buildable that isn't available
RULE:          in the target release.
RULE: - The rust ecosystem currently isn't yet considered stable enough for
RULE:   classic lib dependencies and transitions in main; therefore the
RULE:   expectation for those packages is to vendor (and own/test) all
RULE:   dependencies (except those provided by the rust runtime itself).
RULE:   This implies that all the rules for vendored builds always
RULE:   apply to them. In addition:
RULE:   - The rules and checks for rust based packages are preliminary and might
RULE:     change over time as the ecosytem matures and while
RULE:     processing the first few rust based packages.
RULE:   - It is expected rust builds will use dh-cargo so that a later switch
RULE:     to non vendored dependencies isn't too complex (e.g. it is likely
RULE:     that over time more common libs shall become stable and then archive
RULE:     packages will be used to build).
RULE:   - the rust tooling e.g. dh-cargo can not yet automatically provide

> RULE:     all that we require - for example Cargo.lock - until it does the

Reviewer 2: I'm very concerned that adopting these rules before we have a good `Cargo.lock` story will create a very difficult situation to cope with:

- At some point, we will need to provide SBOMs to sell to specific clients and into specific industries.
- Having detailed summarized information available short of "search all the source code" is a basic necessity for maintaining these packages ten years from now.

While I'm overall a big fan of the "get something working and then iterate on it as needs are identified", I'm also very scared that we'll wind up with dozens of packages and no tooling to track the vendored dependencies.

If we let in one, the arguments against the second, third, and fourth get increasingly weak.

I believe we need a solution to this problem before we let in the first package.

Thanks

RULE:     package build shall be adapted in a way to provide that without
RULE:     dh* tooling support.
RULE: - All vendored dependencies (no matter what language) shall have a
RULE:   way to be refreshed
TODO-A: - This does not use static builds
TODO-B: - The team TBD is aware of the implications by a static build and
TODO-B:   commits to test no-change-rebuilds and to fix any issues found for the
TODO-B:   lifetime of the release (including ESM)

TODO-A: - This does not use vendored code
TODO-B: - The team TBD is aware of the implications of vendored code and (as
TODO-B:   alerted by the security team) commits to provide updates and backports
TODO-B:   to the security team for any affected vendored code for the lifetime
TODO-B:   of the release (including ESM).

TODO-A: - This does not use vendored code

> TODO-B: - This package uses vendored go code tracked in go.mod as shiped in the

Reviewer 3: the direct dependencies are in `go.mod`. Exact checksum content for direct and indirect dependencies are in `go.sum`. Reproducible builds involve having a `go.sum` file.

Author: Thanks, I adopted the section (and one other reference to it) marking resolved

RULE:   to allow static builds of golang packages in main, so long as the number
RULE:   of these packages remains low and they follow the guidelines below:
RULE:   - golang applications in main are expected:

> RULE:       1. to build golang use `golang-*-dev` packages creating
> RULE:          `Built-Using` in debian/control. This requirement ensures

Reviewer 4: I found the old wording more comprehensible.
```suggestion
RULE:       1. to build using `golang-*-dev` packages from the Ubuntu archive
RULE:          creating `Built-Using` in debian/control. This requirement ensures
```

TODO-B:   - golang: static builds are used, the team confirmed their commitment
TODO-B:     to the additional responsibilities implied by static builds.

TODO-A: - not a rust package, no extra constraints to consider in that regard
TODO-B: - Rust package that has all dependencies vendored. It does neither

> TODO-B:   have *Built-Using (after build). Nor does the biuld log indicate

Reviewer 4: ```suggestion
TODO-B:   have *Built-Using (after build). Nor does the build log indicate
```
